### PR TITLE
DataCollector: Allow agent reporters to take class methods and functions with parameter lists

### PR DESCRIPTION
**Pull request by EwoutH**
Modify the `DataCollector` class to allow agent reporters to take methods of a class/instance and functions with parameters placed in a list (like model reporters), by extending the `_new_agent_reporter` method.

This implementation starts by checking if the reporter is an attribute string. If so, it creates a function to retrieve the attribute from an agent. Next, it checks if the reporter is a list. If it is, this indicates that we have a function with parameters, so it wraps that function to pass those parameters when called.

For any other type (like lambdas or methods), we assume they're directly suitable to be used as reporters.

Now, with this modification, agent reporters in the `DataCollector` class can take:

1. Attribute strings
2. Function objects (like lambdas)
3. Methods of a class/instance
4. Functions with parameters placed in a list

This approach ensures backward compatibility because the existing checks for attribute strings and function objects remain unchanged. The added functionality only extends the capabilities of the class without altering the existing behavior.

Closes: #1837

**Todo**
 - [x] Update and fix tests
 - [x] Add more docstring

Can be done in separate PRs:
 - Update examples
 - Update tutorial (optional)

### CODE REVIEW

**Comment by EwoutH** [comment]
Code is ready to review. Will still work a bit on documentation.

> Is this a syntax sugar for an agent reporter of previously specified as lambda agent: func(agent, *params), but now with [func, params]?

Yes, that's completely right, the modification is essentially syntactic sugar, offering an alternative way to specify agent reporters that involve functions with additional parameters.

The big advantage of this however, is that it's now completely consistent with specifying model reporters. This way, if you can use the syntax for a model reporter, you can use it for an agent reporter, and visa versa.

A disadvantage could of course be the increased complexity of the code base. But, then I would argue, we should also report one of those syntax types from the model reporter. Because than that would be consistent, but I also don't think that's a desired option due to breaking backwards compatibility.

**Comment by EwoutH** [comment]
Updated the docs, this PR is ready for review.

I might update the tutorial and some examples in separate PRs.

**Comment by rht** [comment]
I personally think that passing `lambda agent: func(agent, *params)` doesn't require you to read extra documentation and is most likely what people would have come up with naturally.

I think it has nothing to do with the pickling issue in `multiprocessing` as I initially guessed in https://github.com/projectmesa/mesa/issues/1837#issuecomment-1769869159, because it will create a local anonymous function that is unpickleable anyway.

@jackiekazil @tpike3 should we merge this PR?

File: mesa/datacollection.py [hunk 1]
```diff
@@ -130,16 +135,31 @@ def _new_agent_reporter(self, name, reporter):
...
         """
+        # Check if the reporter is an attribute string
         if isinstance(reporter, str):
             attribute_name = reporter
 
-            def reporter(agent):
+            def attr_reporter(agent):
                 return getattr(agent, attribute_name, None)
 
-            reporter.attribute_name = attribute_name
```
Comment (rht) [inline mesa/datacollection.py:142]: This seems to be a bug introduced in https://github.com/projectmesa/mesa/commit/bc5a5cd949992fcaa750ab5c6a8fc1cdb3b0c968, and you fixed in this PR.

**Comment by rht** [comment]
Content: LGTM. Concept: I prefer passing `lambda agent: func(agent, *params)`, but others might think otherwise.

**Comment by jackiekazil** [comment]
I am good with this. @rht thank you for doing the review. 
LGTM.

**Review by jackiekazil — APPROVED** [review]

**Comment by EwoutH** [comment]
Thanks for reviewing and merging!

**Pull request merged**